FirebaseDeviceStore/FirebaseDeviceStore.swift

```diff
@@ -108,90 +141,20 @@ public class FirebaseDeviceStore: NSObject, MessagingDelegate {
-
-            if (doc.exists) {
-                var devices = self.getDevices(doc);
-                // Remove the old device
-                devices = self.removeCurrentDevice(devices);
-                // Update the document
-                transaction.updateData([self.DEVICES_FIELD: devices], forDocument: docRef);
-            } else {
-                let devices = self.createUserDevices(userId, nil);
-                transaction.setData(devices, forDocument: docRef);
-            }
-            return nil;
-        }) { (object, error) in
-            // TODO: Logging
-        }
+        docRef.updateData([FieldPath.init([self.DEVICES_FIELD, self.getDeviceId()]): FieldValue.delete()], completion: completion);
     }
 
-    private func updateDevice(_ userId: String, _ token: String) {
+    private func updateDevice(_ userId: String, _ token: String, completion: @escaping (Error?) -> Void) {
         let docRef = userRef(userId);
-
-        firestore.runTransaction({ (transaction, errorPointer) -> Any? in
-            let doc: DocumentSnapshot;
-            do {
-                try doc = transaction.getDocument(docRef);
-            } catch let fetchError as NSError {
-                errorPointer?.pointee = fetchError
-                return nil;
-            }
-            if (doc.exists) {
-                var devices = self.getDevices(doc);
-                if (self.containsCurrentDevice(devices)) {
-                    // Update the device token if it already exists
-                    self.updateCurrentDevice(devices, token);
-                } else {
-                    devices.append(self.createCurrentDevice(token));
-                }
-                // Update the document
-                transaction.updateData([self.DEVICES_FIELD: devices], forDocument: docRef);
-            } else {
-                let devices = self.createUserDevices(userId, token);
-                transaction.setData(devices, forDocument: docRef);
-            }
-            return nil;
-        }) { (object, error) in
-            // TODO: Logging
-        }
-    }
-
-    private func containsCurrentDevice(_ devices: [[String: String]]) -> Bool {
-        let deviceId = getDeviceId();
-        for device in devices {
-            if (deviceId == device[DEVICE_ID_FIELD]) {
-                return true;
-            }
-        }
-        return false;
+        let deviceId = self.getDeviceId();
+        docRef.setData([USER_ID_FIELD: userId, DEVICES_FIELD: [deviceId: self.createDevice(deviceId, token)]], merge: true, completion: completion);
     }
 
-    private func createCurrentDevice(_ token: String) -> [String: String] {
+    private func createDevice(_ deviceId: String, _ token: String) -> [String: String] {
         let device: [String: String] = [
-            DEVICE_ID_FIELD: getDeviceId(),
+            DEVICE_ID_FIELD: deviceId,
             FCM_TOKEN_FIELD: token,
             NAME_FIELD: getDeviceName(),
             OS_FIELD: getOS(),
```

Comment:
Random thought (feel free to ignore or do in separate PR): I wonder if it'd be useful to store a timestamp in here (perhaps using `FieldValue.serverTimestamp()` to populate)?  Could be useful to clean up stale devices or avoid sending them notifications or something.

Reply:
Yes, this makes sense and is on the TODO list, but I will do as a separate PR